File: mobile/ios/OpenU60/Features/SMS/SMSModels.swift
import Foundation

// MARK: - Tag Enum

enum SMSTag: Int, Sendable {
    case read = 0
    case unread = 1
    case sent = 2
    case failed = 3
    case draft = 4

    var isIncoming: Bool { self == .read || self == .unread }
}

// MARK: - Storage Filter

enum SMSStorageFilter: String, CaseIterable, Sendable {
    case all = "All"
    case `internal` = "Internal"
    case sim = "SIM"

    var memStoreValue: Int {
        switch self {
        case .all: return 2
        case .internal: return 1
        case .sim: return 0
        }
    }
}

// MARK: - SMS Message

struct SMSMessage: Identifiable, Sendable {
    let id: Int
    let number: String
    let content: String
    let date: Date
    let tag: SMSTag
    let groupId: String
    let memStore: String
}

// MARK: - SMS Conversation

struct SMSConversation: Identifiable {
    var id: String { normalizedNumber }
    let normalizedNumber: String
    let number: String
    var messages: [SMSMessage]
    var unreadCount: Int
    var latestMessage: String
    var latestTime: Date
}

// MARK: - SMS Capacity

struct SMSCapacity {
    let nvTotal: Int
    let nvUsed: Int
    let simTotal: Int
    let simUsed: Int
    let unreadCount: Int

    static let empty = SMSCapacity(nvTotal: 0, nvUsed: 0, simTotal: 0, simUsed: 0, unreadCount: 0)
}

// MARK: - Parser

enum SMSParser {

    /// Decode UCS-2 hex string (UTF-16BE, 4 hex chars per character) to readable text.
    static func decodeUCS2Hex(_ hex: String) -> String {
        var result = ""
        let chars = Array(hex)
        var i = 0
        while i + 3 < chars.count {
            let hexStr = String(chars[i...i+3])
            if let scalar = UInt32(hexStr, radix: 16), let unicode = Unicode.Scalar(scalar) {
                result.append(Character(unicode))
            }
            i += 4
        }
        return result
    }

    /// Encode text to UCS-2 hex string (UTF-16BE).
    static func encodeUCS2Hex(_ text: String) -> String {
        text.utf16.map { String(format: "%04X", $0) }.joined()
    }

    /// Parse ZTE date format "YY,MM,DD,HH,MM,SS,TZ" to Date.
    static func parseSMSDate(_ dateStr: String) -> Date {
        let parts = dateStr.split(separator: ",").map(String.init)
        guard parts.count >= 6 else { return Date() }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!

        let year = (Int(parts[0]) ?? 0) + 2000
        let month = Int(parts[1]) ?? 1
        let day = Int(parts[2]) ?? 1
        let hour = Int(parts[3]) ?? 0
        let minute = Int(parts[4]) ?? 0
        let second = Int(parts[5]) ?? 0

        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = day
        components.hour = hour
        components.minute = minute
        components.second = second

        // Parse timezone offset if present (e.g. "+0", "+32" = +8h in quarter-hours)
        if parts.count >= 7 {
            let tzStr = parts[6].trimmingCharacters(in: .whitespaces)
            if let quarters = Int(tzStr) {
                let offsetSeconds = quarters * 15 * 60
                components.timeZone = TimeZone(secondsFromGMT: offsetSeconds)
            }
        }

        return calendar.date(from: components) ?? Date()
    }

    /// Format current time in ZTE SMS send format (semicolons, tz in hours).
    /// JS: "YY;MM;DD;HH;MM;SS;+TZ" where TZ is offset in hours.
    static func formatSMSTime() -> String {
        let now = Date()
        let calendar = Calendar(identifier: .gregorian)
        let tz = TimeZone.current
        let comps = calendar.dateComponents(in: tz, from: now)

        let year = (comps.year ?? 2026) % 100
        let offsetHours = tz.secondsFromGMT() / 3600
        let tzStr = offsetHours >= 0 ? "+\(offsetHours)" : "\(offsetHours)"

        return String(format: "%02d;%02d;%02d;%02d;%02d;%02d;%@",
                      year, comps.month ?? 1, comps.day ?? 1,
                      comps.hour ?? 0, comps.minute ?? 0, comps.second ?? 0, tzStr)
    }

    /// Determine encode type for sending.
    static func getEncodeType(_ text: String) -> String {
        // Check if all characters are in GSM 7-bit default alphabet
        let gsm7 = CharacterSet(charactersIn:
            "@£$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ ÆæßÉ" +
            " !\"#¤%&'()*+,-./0123456789:;<=>?" +
            "¡ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
            "ÄÖÑÜabcdefghijklmnopqrstuvwxyz" +
            "äöñüà§")
        if text.unicodeScalars.allSatisfy({ gsm7.contains($0) }) {
            return "GSM7_default"
        }
        return "UNICODE"
    }

    /// Normalize phone number to last 8 digits for conversation grouping.
    static func normalizeNumber(_ number: String) -> String {
        let digits = number.filter(\.isNumber)
        if digits.count > 8 {
            return String(digits.suffix(8))
        }
        return digits
    }

    /// Group messages into conversations by normalized phone number.
    static func groupIntoConversations(_ messages: [SMSMessage]) -> [SMSConversation] {
        var grouped: [String: [SMSMessage]] = [:]

        for msg in messages {
            let key = normalizeNumber(msg.number)
            grouped[key, default: []].append(msg)
        }

        return grouped.map { key, msgs in
            let sorted = msgs.sorted { $0.date < $1.date }
            let latest = sorted.last!
            let unread = msgs.filter { $0.tag == .unread }.count
            // Use the longest number variant as display number
            let displayNumber = msgs.map(\.number).max(by: { $0.count < $1.count }) ?? latest.number

            return SMSConversation(
                normalizedNumber: key,
                number: displayNumber,
                messages: sorted,
                unreadCount: unread,
                latestMessage: latest.content,
                latestTime: latest.date
            )
        }.sorted { $0.latestTime > $1.latestTime }
    }

    /// Parse raw SMS data from the agent into SMSMessage array.
    /// Response format: { "messages": [ { "id":Int, "number":String, "content":String(UCS2hex),
    ///   "date":"YY,MM,DD,HH,MM,SS,TZ", "tag":"0".."4", "draft_group_id":String, "mem_store":String } ] }
    static func parseMessages(_ data: [String: Any]) -> [SMSMessage] {
        guard let list = data["messages"] as? [[String: Any]] else { return [] }

        return list.compactMap { item -> SMSMessage? in
            guard let id = item["id"] as? Int,
                  let number = item["number"] as? String,
                  let body = item["content"] as? String,
                  let dateStr = item["date"] as? String,
                  let tagStr = item["tag"] as? String,
                  let tagInt = Int(tagStr),
                  let tag = SMSTag(rawValue: tagInt) else { return nil }

            // Content is always UCS-2 hex encoded; decode it.
            // If decoding produces empty string, fall back to raw (might be GSM7 plain text).
            let decoded = decodeUCS2Hex(body)
            let content = decoded.isEmpty ? body : decoded

            return SMSMessage(
                id: id,
                number: number,
                content: content,
                date: parseSMSDate(dateStr),
                tag: tag,
                groupId: item["draft_group_id"] as? String ?? "",
                memStore: item["mem_store"] as? String ?? "nv"
            )
        }
    }

    /// Parse capacity response.
    static func parseCapacity(_ data: [String: Any]) -> SMSCapacity {
        SMSCapacity(
            nvTotal: data["sms_nv_total"] as? Int ?? 0,
            nvUsed: data["sms_nvused_total"] as? Int ?? 0,
            simTotal: data["sms_sim_total"] as? Int ?? 0,
            simUsed: data["sms_simused_total"] as? Int ?? 0,
            unreadCount: data["sms_dev_unread_num"] as? Int ?? 0
        )
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/core/model/SMSModels.kt
package com.openu60.core.model

import java.util.*

// MARK: - Tag Enum

enum class SMSTag(val value: Int) {
    READ(0), UNREAD(1), SENT(2), FAILED(3), DRAFT(4);

    val isIncoming: Boolean get() = this == READ || this == UNREAD

    companion object {
        fun fromInt(v: Int): SMSTag? = entries.firstOrNull { it.value == v }
    }
}

// MARK: - Storage Filter

enum class SMSStorageFilter(val memStoreValue: Int, val label: String) {
    ALL(2, "All"),
    INTERNAL(1, "Internal"),
    SIM(0, "SIM");
}

// MARK: - SMS Message

data class SMSMessage(
    val id: Int,
    val number: String,
    val content: String,
    val date: Date,
    val tag: SMSTag,
    val groupId: String,
    val memStore: String,
)

// MARK: - SMS Conversation

data class SMSConversation(
    val normalizedNumber: String,
    val number: String,
    val messages: List<SMSMessage>,
    val unreadCount: Int,
    val latestMessage: String,
    val latestTime: Date,
) {
    val id: String get() = normalizedNumber
}

// MARK: - SMS Capacity

data class SMSCapacity(
    val nvTotal: Int = 0,
    val nvUsed: Int = 0,
    val simTotal: Int = 0,
    val simUsed: Int = 0,
    val unreadCount: Int = 0,
) {
    companion object {
        val empty = SMSCapacity()
    }
}

// MARK: - Parser

object SMSParser {

    fun decodeUCS2Hex(hex: String): String {
        val sb = StringBuilder()
        val chars = hex.toCharArray()
        var i = 0
        while (i + 3 < chars.size) {
            val hexStr = String(chars, i, 4)
            val code = hexStr.toIntOrNull(16)
            if (code != null && code != 0) {
                sb.append(code.toChar())
            }
            i += 4
        }
        return sb.toString()
    }

    fun encodeUCS2Hex(text: String): String {
        return text.map { "%04X".format(it.code) }.joinToString("")
    }

    fun parseSMSDate(dateStr: String): Date {
        val parts = dateStr.split(",").map { it.trim() }
        if (parts.size < 6) return Date()

        val year = (parts[0].toIntOrNull() ?: 0) + 2000
        val month = (parts[1].toIntOrNull() ?: 1) - 1
        val day = parts[2].toIntOrNull() ?: 1
        val hour = parts[3].toIntOrNull() ?: 0
        val minute = parts[4].toIntOrNull() ?: 0
        val second = parts[5].toIntOrNull() ?: 0

        val cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        cal.set(year, month, day, hour, minute, second)
        cal.set(Calendar.MILLISECOND, 0)

        if (parts.size >= 7) {
            val tzStr = parts[6].trim()
            val quarters = tzStr.toIntOrNull()
            if (quarters != null) {
                val offsetMs = quarters * 15 * 60 * 1000
                cal.timeZone = TimeZone.getTimeZone("GMT")
                cal.add(Calendar.MILLISECOND, -offsetMs)
            }
        }

        return cal.time
    }

    fun formatSMSTime(): String {
        val cal = Calendar.getInstance()
        val year = cal.get(Calendar.YEAR) % 100
        val offsetHours = cal.timeZone.rawOffset / 3600000
        val tzStr = if (offsetHours >= 0) "+$offsetHours" else "$offsetHours"
        return "%02d;%02d;%02d;%02d;%02d;%02d;%s".format(
            year, cal.get(Calendar.MONTH) + 1, cal.get(Calendar.DAY_OF_MONTH),
            cal.get(Calendar.HOUR_OF_DAY), cal.get(Calendar.MINUTE), cal.get(Calendar.SECOND), tzStr,
        )
    }

    fun getEncodeType(text: String): String {
        val gsm7 = "@£\$¥èéùìòÇ\nØø\rÅåΔ_ΦΓΛΩΠΨΣΘΞ ÆæßÉ" +
            " !\"#¤%&'()*+,-./0123456789:;<=>?" +
            "¡ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
            "ÄÖÑÜabcdefghijklmnopqrstuvwxyz" +
            "äöñüà§"
        return if (text.all { it in gsm7 }) "GSM7_default" else "UNICODE"
    }

    fun normalizeNumber(number: String): String {
        val digits = number.filter { it.isDigit() }
        return if (digits.length > 8) digits.takeLast(8) else digits
    }

    fun groupIntoConversations(messages: List<SMSMessage>): List<SMSConversation> {
        val grouped = mutableMapOf<String, MutableList<SMSMessage>>()
        for (msg in messages) {
            val key = normalizeNumber(msg.number)
            grouped.getOrPut(key) { mutableListOf() }.add(msg)
        }
        return grouped.map { (key, msgs) ->
            val sorted = msgs.sortedBy { it.date }
            val latest = sorted.last()
            val unread = msgs.count { it.tag == SMSTag.UNREAD }
            val displayNumber = msgs.maxByOrNull { it.number.length }?.number ?: latest.number
            SMSConversation(
                normalizedNumber = key,
                number = displayNumber,
                messages = sorted,
                unreadCount = unread,
                latestMessage = latest.content,
                latestTime = latest.date,
            )
        }.sortedByDescending { it.latestTime }
    }

    fun parseMessages(data: Map<String, Any?>): List<SMSMessage> {
        val list = data["messages"] as? List<*> ?: return emptyList()
        return list.mapNotNull { item ->
            val map = item as? Map<*, *> ?: return@mapNotNull null
            val id = DeviceParser.asInt(map["id"]) ?: return@mapNotNull null
            val number = map["number"] as? String ?: return@mapNotNull null
            val body = map["content"] as? String ?: return@mapNotNull null
            val dateStr = map["date"] as? String ?: return@mapNotNull null
            val tagStr = map["tag"] as? String ?: return@mapNotNull null
            val tagInt = tagStr.toIntOrNull() ?: return@mapNotNull null
            val tag = SMSTag.fromInt(tagInt) ?: return@mapNotNull null

            val decoded = decodeUCS2Hex(body)
            val content = if (decoded.isEmpty()) body else decoded

            SMSMessage(
                id = id,
                number = number,
                content = content,
                date = parseSMSDate(dateStr),
                tag = tag,
                groupId = map["draft_group_id"] as? String ?: "",
                memStore = map["mem_store"] as? String ?: "nv",
            )
        }
    }

    fun parseCapacity(data: Map<String, Any?>): SMSCapacity {
        return SMSCapacity(
            nvTotal = DeviceParser.asInt(data["sms_nv_total"]) ?: 0,
            nvUsed = DeviceParser.asInt(data["sms_nvused_total"]) ?: 0,
            simTotal = DeviceParser.asInt(data["sms_sim_total"]) ?: 0,
            simUsed = DeviceParser.asInt(data["sms_simused_total"]) ?: 0,
            unreadCount = DeviceParser.asInt(data["sms_dev_unread_num"]) ?: 0,
        )
    }
}
